pyserini/dsearch/__main__.py

```diff
@@ -106,11 +112,27 @@ def init_query_encoder(encoder, topics_name, device):
                 topic_batch = [topics[topic].get('title').strip() for topic in topic_key_batch]
                 hits = searcher.batch_search(topic_batch, topic_key_batch, k=args.hits, threads=args.threads)
                 for topic in hits:
+                    unique_docs = set()
+                    rank = 1
                     for idx, hit in enumerate(hits[topic]):
-                        if args.msmarco:
-                            target_file.write(f'{topic}\t{hit.docid}\t{idx + 1}\n')
+                        if args.max_passage:
+                            docid, _ = hit.docid.split(args.max_passage_delimiter)
+                            if docid in unique_docs:
+                                continue
+                            if args.msmarco:
```

Comment:
I noticed this conditional: if x write this way, else write that way... is duplicated in several places. This is potentially dangerous if we change something, leading to inconsistent code paths.

Is there an obvious clean way to fix this?

Reply:
should be able to clean, like what we have done in `search.__main__.py`. But I prefer to that in a separate PR, together with the hybrid document search. `hsearch.__main__.py` will have similar things to deal with so I feel it's good to be cleaned together.

Reply:
sg